src/pyqasm/transformer.py

```diff
@@ -438,3 +440,142 @@ def get_type_string(variable: Variable) -> str:
+        qubit_register_offsets: dict[str, int],
+        global_qreg_size_map: dict[str, int],
+        device_qubits: int | None,
+    ) -> Sequence[Statement] | Statement:
+        """Transform statements by mapping qubit registers to device qubit register indices
+
+        Args:
+            unrolled_stmts : The statements or single statement to transform.
+            qubit_register_offsets (dict): Mapping from register name to its
+                                           offset in the global qubit array.
+            global_qreg_size_map (dict): original global qubit register mapping.
+            device_qubits (int): Total number of device qubits
+
+        Returns:
+            The transformed statements or statement with qubit registers mapped to device indices.
+        """
+        if device_qubits is None:
+            device_qubits = sum(global_qreg_size_map.values())
+
+        def _get_pyqasm_device_qubit_index(
+            reg: str, idx: int, qubit_reg_offsets: dict[str, int], global_qreg: dict[str, int]
+        ):
+            _offsets = qubit_reg_offsets
+            _n_qubits = global_qreg[reg]
+            if not 0 <= idx < _n_qubits:
+                raise IndexError(f"{reg}[{idx}] out of range (0..{_n_qubits-1})")
+            return _offsets[reg] + idx
+
+        if isinstance(unrolled_stmts, QuantumBarrier):
+            _qubit_id = cast(Identifier, unrolled_stmts.qubits[0])  # type: ignore[union-attr]
+            if not isinstance(_qubit_id, IndexedIdentifier):
+                _start = _get_pyqasm_device_qubit_index(
+                    _qubit_id.name, 0, qubit_register_offsets, global_qreg_size_map
+                )
+                _end = _get_pyqasm_device_qubit_index(
+                    _qubit_id.name,
+                    global_qreg_size_map[_qubit_id.name] - 1,
+                    qubit_register_offsets,
+                    global_qreg_size_map,
+                )
+                if _start == 0:
+                    _qubit_id.name = f"__PYQASM_QUBITS__[:{_end+1}]"
+                elif _end == device_qubits - 1:
+                    _qubit_id.name = f"__PYQASM_QUBITS__[{_start}:]"
+                else:
+                    _qubit_id.name = f"__PYQASM_QUBITS__[{_start}:{_end+1}]"
+            else:
+                _qubit_str = cast(str, unrolled_stmts.qubits[0].name)  # type: ignore[union-attr]
+                _qubit_ind = cast(
+                    list, unrolled_stmts.qubits[0].indices
+                )  # type: ignore[union-attr]
+                for multi_ind in _qubit_ind:
+                    for ind in multi_ind:
+                        pyqasm_ind = _get_pyqasm_device_qubit_index(
+                            _qubit_str.name, ind.value, qubit_register_offsets, global_qreg_size_map
+                        )
+                        ind.value = pyqasm_ind
+                _qubit_str.name = "__PYQASM_QUBITS__"
+            return unrolled_stmts
```

Comment:
can move this out of the checks, because you are essentially doing this for every case